R/dictionary_based_cleaning.R

```diff
@@ -0,0 +1,228 @@
+  stopifnot(all(c("options", "values", "grp") %in% names(dictionary)),
+            all(unique(dictionary[["grp"]]) %in% names(data)))
+
+  # detect misspelled options in the columns to clean
+  misspelled_options <- detect_misspelled_options(data, dictionary)
+
+  # correct the misspelled options if they exist
+  if (length(misspelled_options) > 0L) {
+    if (correct) {
+      res    <- correct_misspelled_options(data, dictionary, misspelled_options)
+      data   <- add_report(res[["data"]],
+                           res[["report"]],
+                           name = "misspel_correction")
+
+      # perform the dictionary-based cleaning
+      data   <- matchmaker::match_df(data,
+                                     dictionary = dictionary,
+                                     from       = "options",
+                                     to         = "values",
+                                     by         = "grp")
+      data   <- add_report(data,
+                           unique(dictionary[["grp"]]),
+                           name = "columns_modified_based_on_dictionary")
+
+    } else {
+      print_misspelled_values(misspelled_options)
+      message("Please correct them first or use `correct=TRUE` for ",
+              "auto-correction")
+      return(NULL)
+    }
+  }
+
+  data
+}
+
+#' Detect misspelled options in columns to be cleaned
+#'
+#' @inheritParams clean_using_dictionary
+#'
+#' @return a `list` with the indexes of the misspelled values in every column
+#'    that needs to be cleaned
+#' @keywords internal
+#' @noRd
+#'
+detect_misspelled_options <- function(data, dictionary) {
+  grp <- NULL
+  cols_to_modify  <- unique(dictionary[["grp"]])
+  outliers        <- list()
+  for (col in cols_to_modify) {
+    unique_values <- unique(data[[col]])[!is.na(unique(data[[col]]))]
+    temp_dict     <- dictionary %>%
+      dplyr::filter(grp == col)
+    opts          <- c(temp_dict[["options"]], unique(temp_dict[["values"]]))
+    m             <- match(unique_values, opts)
+    which(!(unique(data[[col]]) %in% dictionary[["options"]]))
+    if (anyNA(m)) {
+      outliers[[col]] <- which(data[[col]] == unique_values[which(is.na(m))])
+    }
+  }
+  outliers
+}
+
+#' Correct the misspelled values with their closest values from the data
+#' dictionary
+#'
+#' @inheritParams clean_using_dictionary
+#' @param outliers
+#'
+#' @importFrom stringdist stringdist
+#' @keywords internal
+#' @noRd
+correct_misspelled_options <- function(data, dictionary, outliers) {
+  grp <- NULL
+  checkmate::assert_list(outliers, min.len = 1L, null.ok = FALSE)
+  for (target_col in names(outliers)) {
+    target_values    <- dictionary %>%
+      dplyr::filter(grp == target_col)
+    target_values    <- unique(target_values[["values"]])
+    misspelled_words <- data[[target_col]][outliers[[target_col]]]
+    target_dist      <- do.call("rbind", lapply(misspelled_words,
+                                                stringdist::stringdist,
```

Comment:
The autocorrect enhancement comes at the cost of an additional package `{stringdist}` which is a big price to pay for a niche feature that can be resolved by users adding extra options to their matchmaker dictionary list. 

I think this function can be replaced with one that allows the user to easily add records to the dictionary.

Reply:
I have removed the function designed to automatically correct for the misspelled options. This is now replaced with a fucntion to add an option to the dictionary.